### DISCUSSION TITLE
vitessce example with scrnaseq and image

### DISCUSSION THREAD

**Discussion by FADHLyemen** [vitessce/vitessce-python#95] [Q&A] [answered]
Hi Vitessce team, 
I am interested to have one example combining scrna-seq and image. 
I could not find it in 
https://github.com/vitessce/vitessce-python/tree/master/docs/notebooks

I want to have something like:
http://vitessce.io/?dataset=linnarsson-2018&theme=dark

I am wondering on how to connect the h5ad with the image coordinates.

Thank you

**Comment by keller-mark** [comment]
Hi @FADHLyemen,
There is a script here which demonstrates processing of an AnnData object containing an image file: https://github.com/vitessce/vitessce-python/blob/keller-mark/demo-data-processing/demos/human_lymph_node_10x_visium/src/create_zarr.py

It is not very well documented but may be helpful for you. The corresponding Vitessce configuration is here as raw JSON https://github.com/vitessce/vitessce/blob/8b07822f06da5e6d57d381c0633e7cd471cfbc7c/src/app/api.js#L77. We need to work on making a notebook which shows how to do configure it from Python.

**Reply by ilan-gold** [reply]
While the data is not public, here is an example of setting up a view configuration for a segmentation mask + image + `AnnData` container with quantification and clustering:

https://github.com/hubmapconsortium/portal-ui/blob/846d2b326af5e8eb46a5ef449bb3d651acdc64f5/context/app/api/vitessce_confs/base_confs.py#L363-L406

**Reply by ilan-gold** [reply]
This is for the [HuBMAP data portal](portal.hubmapconsortium.org) but the data itself is not public yet.

The main differences between this and non-spatial data is the addition of "spatial" keys to the `AnnDataWrapper` like:
https://github.com/hubmapconsortium/portal-ui/blob/main/context/app/api/vitessce_confs/base_confs.py#L377

and the presence of pyramidal OME-TIFF images + segmentation files:
https://github.com/hubmapconsortium/portal-ui/blob/846d2b326af5e8eb46a5ef449bb3d651acdc64f5/context/app/api/vitessce_confs/base_confs.py#L398-L402

and then of course a `SPATIAL` component:
https://github.com/hubmapconsortium/portal-ui/blob/846d2b326af5e8eb46a5ef449bb3d651acdc64f5/context/app/api/vitessce_confs/base_confs.py#L409

**Comment by FADHLyemen** [comment]
Thank you, 
for the demo, 
http://beta.vitessce.io/index.html?dataset=codeluppi-2018

could I download the data and run it locally using vitessce python?
Can I use any of your fake data to run it in my local vitessce or uplaod them to the s3 aws bucket?

**Comment by FADHLyemen** [comment]
for example for this demo
http://vitessce.io/?dataset=linnarsson-2018&theme=dark
the fake data is in:
https://github.com/vitessce/vitessce-data/tree/master/fake-files/input/linnarsson

but it has two h5ad files
I need to import them and do the same pipeline as in 
https://github.com/vitessce/vitessce-python/blob/master/docs/notebooks/data_export_s3.ipynb

**Reply by ilan-gold** [reply]
@FADHLyemen The two files in there are `hdf5` which are different than `h5ad` - a superset relationship, to be exact.  `h5ad` is a wrapper around `hdf5` files for saving [`AnnData`](https://anndata.readthedocs.io/en/latest/) objects to disk from their python representation.  So all `h5ad` files are `hdf5` but not vice-versa.  Relatedly, `AnnData` can also be stored as [`zarr`](https://zarr.readthedocs.io/en/stable/) via the `write_zarr` method on an `AnnData` object.

**Reply by FADHLyemen** [reply]
@ilan-gold thank you for this. 
so can I use this these two fake hdf5 to generate something in your demo:
http://vitessce.io/?dataset=linnarsson-2018&theme=dark

Maybe an easy way is to put my jason files in amazon s3 and paste them in vitessce app. maybe, we do not need the notebook example. I am trying to replicate your demo, so I can just replace your fake files with my similar files.



```
{
  "name": "Codeluppi et al., Nature Methods 2018",
  "description": "Spatial organization of the somatosensory cortex revealed by osmFISH",
  "version": "1.0.0",
  "initStrategy": "auto",
  "datasets": [
    {
      "uid": "codeluppi",
      "name": "Codeluppi",
      "files": [
        {
          "type": "cells",
          "fileType": "cells.json",
          "url": "https://s3.amazonaws.com/vitessce-data/0.0.31/master_release/linnarsson/linnarsson.cells.json"
        },
        {
          "type": "cell-sets",
          "fileType": "cell-sets.json",
          "url": "https://s3.amazonaws.com/vitessce-data/0.0.31/master_release/linnarsson/linnarsson.cell-sets.json"
        },
        {
          "type": "raster",
          "fileType": "raster.json",
          "url": "https://s3.amazonaws.com/vitessce-data/0.0.31/master_release/linnarsson/linnarsson.raster.json"
        },
        {
          "type": "molecules",
          "fileType": "molecules.json",
          "url": "https://s3.amazonaws.com/vitessce-data/0.0.31/master_release/linnarsson/linnarsson.molecules.json"
        },
        {
          "type": "neighborhoods",
          "fileType": "neighborhoods.json",
          "url": "https://s3.amazonaws.com/vitessce-data/0.0.31/master_release/linnarsson/linnarsson.neighborhoods.json"
        },
        {
          "type": "expression-matrix",
          "fileType": "clusters.json",
          "url": "https://s3.amazonaws.com/vitessce-data/0.0.31/master_release/linnarsson/linnarsson.clusters.json"
        }
      ]
    }
  ],
  "public": true,
  "coordinationSpace": {
    "embeddingZoom": {
      "PCA": 0,
      "TSNE": 0.75
    },
    "embeddingType": {
      "PCA": "PCA",
      "TSNE": "t-SNE"
    },
    "spatialZoom": {
      "A": -5.5
    },
    "spatialTargetX": {
      "A": 16000
    },
    "spatialTargetY": {
      "A": 20000
    }
  },
  "layout": [
    {
      "component": "description",
      "props": {
        "description": "Codeluppi et al., Nature Methods 2018: Spatial organization of the somatosensory cortex revealed by osmFISH"
      },
      "x": 0,
      "y": 0,
      "w": 2,
      "h": 1
    },
    {
      "component": "layerController",
      "x": 0,
      "y": 1,
      "w": 2,
      "h": 4
    },
    {
      "component": "status",
      "x": 0,
      "y": 5,
      "w": 2,
      "h": 1
    },
    {
      "component": "spatial",
      "coordinationScopes": {
        "spatialZoom": "A",
        "spatialTargetX": "A",
        "spatialTargetY": "A"
      },
      "x": 2,
      "y": 0,
      "w": 4,
      "h": 4
    },
    {
      "component": "genes",
      "x": 9,
      "y": 0,
      "w": 3,
      "h": 2
    },
    {
      "component": "cellSets",
      "x": 9,
      "y": 3,
      "w": 3,
      "h": 2
    },
    {
      "component": "heatmap",
      "props": {
        "transpose": true
      },
      "x": 2,
      "y": 4,
      "w": 5,
      "h": 2
    },
    {
      "component": "cellSetExpression",
      "x": 7,
      "y": 4,
      "w": 5,
      "h": 2
    },
    {
      "component": "scatterplot",
      "coordinationScopes": {
        "embeddingType": "PCA",
        "embeddingZoom": "PCA"
      },
      "x": 6,
      "y": 0,
      "w": 3,
      "h": 2
    },
    {
      "component": "scatterplot",
      "coordinationScopes": {
        "embeddingType": "TSNE",
        "embeddingZoom": "TSNE"
      },
      "x": 6,
      "y": 2,
      "w": 3,
      "h": 2
    }
  ]
}

```

**Reply by ilan-gold** [reply]
@FADHLyemen Those files are fake files in that they contain much fewer data points/data than the actual demo.  If you want to replicate the demo, it's probably best to actually run the scripts in `vitessce-data` to generate the data, then you can proceed as you said, putting the generated data on the cloud and then replacing the URL's in the view configuration.

**Comment by FADHLyemen** [comment]
@ilan-gold 
when you said, run the scripts, do you mean this [https://github.com/vitessce/vitessce-data/blob/master/scripts/process_linnarsson.sh](url)

second, I already have the expected files as JSON format:
https://github.com/vitessce/vitessce-data/tree/master/fake-files/output-expected/linnarsson

I am trying to think how to replicate what I did with you in
https://github.com/vitessce/vitessce-python/blob/master/docs/notebooks/data_export_s3.ipynb

and which code I should replace the below one to run these json files?


```
vc = VitessceConfig(name='Habib et al', description='COVID-19 Healthy Donor Brain')
dataset = vc.add_dataset(name='Brain').add_object(AnnDataWrapper(
        adata,
        mappings_obsm=["X_umap"],
        mappings_obsm_names=["UMAP"],
        cell_set_obs=["CellType"],
        cell_set_obs_names=["Cell Type"],
        expression_matrix="X",
        matrix_gene_var_filter="top_highly_variable"
))
scatterplot = vc.add_view(dataset, cm.SCATTERPLOT, mapping="UMAP")
cell_sets = vc.add_view(dataset, cm.CELL_SETS)
genes = vc.add_view(dataset, cm.GENES)
heatmap = vc.add_view(dataset, cm.HEATMAP)
vc.layout((scatterplot | (cell_sets / genes)) / heatmap);
```
```
s3 = boto3.resource(
    service_name='s3',
    aws_access_key_id=os.environ['VITESSCE_S3_ACCESS_KEY_ID'],
    aws_secret_access_key=os.environ['VITESSCE_S3_SECRET_ACCESS_KEY'],
)

```

**Reply by keller-mark** [reply]
If you have files in the JSON formats supported by Vitessce, you can replace the `.add_object()` function with `.add_file()` (see documentation [here](https://vitessce.github.io/vitessce-python/api_config.html#vitessce.config.VitessceConfigDataset.add_file)).

Like this:

```python3
from vitessce import VitessceConfig, DataType as dt, FileType as ft

vc = VitessceConfig(name='', description='')
dataset = (
    vc.add_dataset(name='Brain')
    .add_file(
        url="http://example.com/cells.json",
        data_type=dt.CELLS,
        file_type=ft.CELLS_JSON,
    )
    .add_file(
        url="http://example.com/molecules.json",
        data_type=dt.MOLECULES,
        file_type=ft.MOLECULES_JSON,
    )
)
```

**Reply by keller-mark** [reply]
The data types and file types are defined here: https://vitessce.github.io/vitessce-python/api_config.html#vitessce.constants.DataType

**Reply by keller-mark** [reply]
Note that data added to the configuration via `.add_file` is not included in the upload when running the `vc.export()` function.

**Comment by FADHLyemen** [comment, marked as answer]
@ilan-gold 
what is file type for expression matrix

"type": "expression-matrix",
          "fileType": "clusters.json",

https://github.com/vitessce/vitessce-data/tree/master/fake-files/output-expected/linnarsson
I try this code but still not working,


```
from vitessce import VitessceConfig, DataType as dt, FileType as ft

vc = VitessceConfig(name='linnarsson', description='Spatial organization of the somatosensory cortex revealed by osmFISH')
dataset = (
    vc.add_dataset(name='linnarsson')
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.cells.json",
        data_type=dt.CELLS,
        file_type=ft.CELLS_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.molecules.json",
        data_type=dt.MOLECULES,
        file_type=ft.MOLECULES_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.cell-sets.json",
        data_type=dt.CELL_SETS,
        file_type=ft.CELL_SETS_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.raster.json",
        data_type=dt.RASTER,
        file_type=ft.RASTER_JSON,
    )
     .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.neighborhoods.json",
        data_type=dt.NEIGHBORHOODS,
        file_type=ft.NEIGHBORHOODS_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.clusters.json",
        data_type=dt.EXPRESSION_MATRIX,
        file_type=ft.ANNDATA_CELLS_ZARR,
    )
)

scatterplot = vc.add_view(dataset, cm.SCATTERPLOT, mapping="t-SNE")
cell_sets = vc.add_view(dataset, cm.CELL_SETS)
genes = vc.add_view(dataset, cm.GENES)
heatmap = vc.add_view(dataset, cm.HEATMAP)
vc.layout((scatterplot | (cell_sets / genes)) / heatmap);


s3 = boto3.resource(
    service_name='s3',
    aws_access_key_id='Advdfvdfvdfvd',
    aws_secret_access_key='vfvfddddddddddddddddddddddddddddddddd',
)

config_dict = vc.export(to='S3', s3=s3, bucket_name='vitejson')


vitessce_url = "http://vitessce.io/?url=data:," + quote_plus(json.dumps(config_dict))
vitessce_url
```

http://vitessce.io/?url=data:,%7B%22version%22%3A+%221.0.2%22%2C+%22name%22%3A+%22linnarsson%22%2C+%22description%22%3A+%22Spatial+organization+of+the+somatosensory+cortex+revealed+by+osmFISH%22%2C+%22datasets%22%3A+%5B%7B%22uid%22%3A+%22A%22%2C+%22name%22%3A+%22linnarsson%22%2C+%22files%22%3A+%5B%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.cells.json%22%2C+%22type%22%3A+%22cells%22%2C+%22fileType%22%3A+%22cells.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.molecules.json%22%2C+%22type%22%3A+%22molecules%22%2C+%22fileType%22%3A+%22molecules.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.cell-sets.json%22%2C+%22type%22%3A+%22cell-sets%22%2C+%22fileType%22%3A+%22cell-sets.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.raster.json%22%2C+%22type%22%3A+%22raster%22%2C+%22fileType%22%3A+%22raster.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.neighborhoods.json%22%2C+%22type%22%3A+%22neighborhoods%22%2C+%22fileType%22%3A+%22neighborhoods.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.clusters.json%22%2C+%22type%22%3A+%22expression-matrix%22%2C+%22fileType%22%3A+%22anndata-cells.zarr%22%7D%5D%7D%5D%2C+%22coordinationSpace%22%3A+%7B%22dataset%22%3A+%7B%22A%22%3A+%22A%22%7D%2C+%22embeddingType%22%3A+%7B%22A%22%3A+%22t-SNE%22%7D%7D%2C+%22layout%22%3A+%5B%7B%22component%22%3A+%22scatterplot%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%2C+%22embeddingType%22%3A+%22A%22%7D%2C+%22x%22%3A+0.0%2C+%22y%22%3A+0.0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+6.0%7D%2C+%7B%22component%22%3A+%22cellSets%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+6.0%2C+%22y%22%3A+0.0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+3.0%7D%2C+%7B%22component%22%3A+%22genes%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+6.0%2C+%22y%22%3A+3.0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+3.0%7D%2C+%7B%22component%22%3A+%22heatmap%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+0%2C+%22y%22%3A+6.0%2C+%22w%22%3A+12%2C+%22h%22%3A+6.0%7D%5D%2C+%22initStrategy%22%3A+%22auto%22%7D

**Reply by FADHLyemen** [reply]
@ilan-gold I appreciate your feedback?
If the above code works to run Vitessce from json files stored on s3 bucket that means our data should also work.

**Reply by ilan-gold** [reply]
> @ilan-gold
> what is file type for expression matrix
> 
> "type": "expression-matrix",
> "fileType": "clusters.json",
> 
> https://github.com/vitessce/vitessce-data/tree/master/fake-files/output-expected/linnarsson

That file type is correct - I'm not really sure why that was called that but that's what it is now.  I'll open a PR to introduce a `expression-matrix.json` to make this clearer.

The issue with your code, I think, seems to be this:

> file_type=ft.ANNDATA_CELLS_ZARR,

It should be `file_type="clusters.json"` if I am not mistaken.

**Reply by ilan-gold** [reply]
Another issue I noticed is that there seems to be a new s3 bucket you are using - I think you will need to set up the permissions again properly for that.  Otherwise, you can re-use your old bucket which has the correct permissions.

**Reply by FADHLyemen** [reply]
@ilan-gold 
here is the problem
there is nothing called clusters. so there is no filetype called clusters.json

![image](https://user-images.githubusercontent.com/23288387/127889007-5f78ab14-a59b-4e7e-a8a3-1d54b62b2c41.png)

for the 3 bucket, I created new one with same old setting.

**Reply by ilan-gold** [reply]
> there is nothing called clusters. so there is no filetype called clusters.json

I understand the issue - those constants are just strings so I suggest setting `file_type="clusters.json"` in your python code.  In the meantime I will get in a constant this in the `ft`.

**Reply by FADHLyemen** [reply]
still not working, 
BTW: I uploaded these files to the s3 bucket:
https://github.com/vitessce/vitessce-data/tree/master/fake-files/output-expected/linnarsson


http://vitessce.io/?url=data:,%7B%22version%22%3A+%221.0.2%22%2C+%22name%22%3A+%22linnarsson%22%2C+%22description%22%3A+%22Spatial+organization+of+the+somatosensory+cortex+revealed+by+osmFISH%22%2C+%22datasets%22%3A+%5B%7B%22uid%22%3A+%22A%22%2C+%22name%22%3A+%22linnarsson%22%2C+%22files%22%3A+%5B%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.cells.json%22%2C+%22type%22%3A+%22cells%22%2C+%22fileType%22%3A+%22cells.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.molecules.json%22%2C+%22type%22%3A+%22molecules%22%2C+%22fileType%22%3A+%22molecules.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.cell-sets.json%22%2C+%22type%22%3A+%22cell-sets%22%2C+%22fileType%22%3A+%22cell-sets.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.raster.json%22%2C+%22type%22%3A+%22raster%22%2C+%22fileType%22%3A+%22raster.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.neighborhoods.json%22%2C+%22type%22%3A+%22neighborhoods%22%2C+%22fileType%22%3A+%22neighborhoods.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.clusters.json%22%2C+%22type%22%3A+%22expression-matrix%22%2C+%22fileType%22%3A+%22clusters.json%22%7D%5D%7D%5D%2C+%22coordinationSpace%22%3A+%7B%22dataset%22%3A+%7B%22A%22%3A+%22A%22%7D%2C+%22embeddingType%22%3A+%7B%22A%22%3A+%22t-SNE%22%7D%7D%2C+%22layout%22%3A+%5B%7B%22component%22%3A+%22scatterplot%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%2C+%22embeddingType%22%3A+%22A%22%7D%2C+%22x%22%3A+0.0%2C+%22y%22%3A+0.0%2C+%22w%22%3A+3.0%2C+%22h%22%3A+6.0%7D%2C+%7B%22component%22%3A+%22cellSets%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+3.0%2C+%22y%22%3A+0.0%2C+%22w%22%3A+3.0%2C+%22h%22%3A+3.0%7D%2C+%7B%22component%22%3A+%22genes%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+3.0%2C+%22y%22%3A+3.0%2C+%22w%22%3A+3.0%2C+%22h%22%3A+3.0%7D%2C+%7B%22component%22%3A+%22heatmap%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+0.0%2C+%22y%22%3A+6.0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+6.0%7D%2C+%7B%22component%22%3A+%22spatial%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+6.0%2C+%22y%22%3A+0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+12%7D%5D%2C+%22initStrategy%22%3A+%22auto%22%7D

![image](https://user-images.githubusercontent.com/23288387/127891632-19b358b4-376a-469a-9bba-8a3fa38c388f.png)

```
from vitessce import VitessceConfig, DataType as dt, FileType as ft

vc = VitessceConfig(name='linnarsson', description='Spatial organization of the somatosensory cortex revealed by osmFISH')
dataset = (
    vc.add_dataset(name='linnarsson')
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.cells.json",
        data_type=dt.CELLS,
        file_type=ft.CELLS_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.molecules.json",
        data_type=dt.MOLECULES,
        file_type=ft.MOLECULES_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.cell-sets.json",
        data_type=dt.CELL_SETS,
        file_type=ft.CELL_SETS_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.raster.json",
        data_type=dt.RASTER,
        file_type=ft.RASTER_JSON,
    )
     .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.neighborhoods.json",
        data_type=dt.NEIGHBORHOODS,
        file_type=ft.NEIGHBORHOODS_JSON,
    )
    .add_file(
        url="https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.clusters.json",
        data_type=dt.EXPRESSION_MATRIX,
        file_type="clusters.json"
    )
)

scatterplot = vc.add_view(dataset, cm.SCATTERPLOT, mapping="t-SNE")
cell_sets = vc.add_view(dataset, cm.CELL_SETS)
genes = vc.add_view(dataset, cm.GENES)
heatmap = vc.add_view(dataset, cm.HEATMAP)
spatial = vc.add_view(dataset, cm.SPATIAL)
vc.layout((scatterplot | (cell_sets / genes)) / heatmap|spatial );


s3 = boto3.resource(
    service_name='s3',
    aws_access_key_id='ggfgdfgdfgsdfgfT',
    aws_secret_access_key='Ggfdgsdfgsdfgdsfgdfs',
)

config_dict = vc.export(to='S3', s3=s3, bucket_name='vitejson')


vitessce_url = "http://vitessce.io/?url=data:," + quote_plus(json.dumps(config_dict))

```

**Reply by ilan-gold** [reply]
@FADHLyemen If you open your browser console and go to the "network" tab, it appears the bucket is misconfigured as all of the data is returning "403 Forbidden" messages.

**Reply by FADHLyemen** [reply]
I changed the policy for this bucket, now I got "failed to load response data'
![image](https://user-images.githubusercontent.com/23288387/127895503-3900525e-ec82-4204-afcc-fc6786986da1.png)

**Reply by FADHLyemen** [reply]
All data is uploaded to the bucket
![image](https://user-images.githubusercontent.com/23288387/127895674-6a283d49-36a9-4a47-a5b3-d4cadc134187.png)

**Reply by FADHLyemen** [reply]
I can easy access the json files from the links:
https://vitejson.s3.us-east-2.amazonaws.com/linnarsson.cells.json
![image](https://user-images.githubusercontent.com/23288387/127897198-c50f6dd7-7dc1-4825-88fb-5206f9454b40.png)

**Reply by FADHLyemen** [reply]
http://vitessce.io/?url=data:,%7B%22version%22%3A+%221.0.2%22%2C+%22name%22%3A+%22linnarsson%22%2C+%22description%22%3A+%22Spatial+organization+of+the+somatosensory+cortex+revealed+by+osmFISH%22%2C+%22datasets%22%3A+%5B%7B%22uid%22%3A+%22A%22%2C+%22name%22%3A+%22linnarsson%22%2C+%22files%22%3A+%5B%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.cells.json%22%2C+%22type%22%3A+%22cells%22%2C+%22fileType%22%3A+%22cells.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.molecules.json%22%2C+%22type%22%3A+%22molecules%22%2C+%22fileType%22%3A+%22molecules.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.cell-sets.json%22%2C+%22type%22%3A+%22cell-sets%22%2C+%22fileType%22%3A+%22cell-sets.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.raster.json%22%2C+%22type%22%3A+%22raster%22%2C+%22fileType%22%3A+%22raster.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.neighborhoods.json%22%2C+%22type%22%3A+%22neighborhoods%22%2C+%22fileType%22%3A+%22neighborhoods.json%22%7D%2C+%7B%22url%22%3A+%22https%3A%2F%2Fvitejson.s3.us-east-2.amazonaws.com%2Flinnarsson.clusters.json%22%2C+%22type%22%3A+%22expression-matrix%22%2C+%22fileType%22%3A+%22clusters.json%22%7D%5D%7D%5D%2C+%22coordinationSpace%22%3A+%7B%22dataset%22%3A+%7B%22A%22%3A+%22A%22%7D%2C+%22embeddingType%22%3A+%7B%22A%22%3A+%22t-SNE%22%7D%7D%2C+%22layout%22%3A+%5B%7B%22component%22%3A+%22scatterplot%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%2C+%22embeddingType%22%3A+%22A%22%7D%2C+%22x%22%3A+0.0%2C+%22y%22%3A+0.0%2C+%22w%22%3A+3.0%2C+%22h%22%3A+6.0%7D%2C+%7B%22component%22%3A+%22cellSets%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+3.0%2C+%22y%22%3A+0.0%2C+%22w%22%3A+3.0%2C+%22h%22%3A+3.0%7D%2C+%7B%22component%22%3A+%22genes%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+3.0%2C+%22y%22%3A+3.0%2C+%22w%22%3A+3.0%2C+%22h%22%3A+3.0%7D%2C+%7B%22component%22%3A+%22heatmap%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+0.0%2C+%22y%22%3A+6.0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+6.0%7D%2C+%7B%22component%22%3A+%22spatial%22%2C+%22coordinationScopes%22%3A+%7B%22dataset%22%3A+%22A%22%7D%2C+%22x%22%3A+6.0%2C+%22y%22%3A+0%2C+%22w%22%3A+6.0%2C+%22h%22%3A+12%7D%5D%2C+%22initStrategy%22%3A+%22auto%22%7D

**Reply by ilan-gold** [reply]
@FADHLyemen Accessing the data as a URL is the same as downloading.  It seems like CORS are misconfigured, if I had to guess.  Please see [this](http://beta.vitessce.io/docs/data-hosting/index.html#cross-origin-resource-sharing-cors) to be sure.

**Reply by FADHLyemen** [reply]
@ilan-gold 
It finally works after adding the below lines, thank you for your patient @ilan-gold 

```
[
  {
    "AllowedHeaders": ["*"],
    "AllowedMethods": ["GET"],
    "AllowedOrigins": ["*"],
    "ExposeHeaders": ["Content-Range"],
    "MaxAgeSeconds": 3000
  }
]
```

**Reply by ilan-gold** [reply]
No Problem!
